<a href="https://colab.research.google.com/github/soujanya-vattikolla/Machine-Learning-Tutorial/blob/main/DecisionTree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Build decision tree model to predict survival based on certain parameters

In [43]:
import pandas as pd

In [44]:
titantic_df = pd.read_csv('titantic.csv')
titantic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [45]:
X = titantic_df[['Pclass', 'Sex', 'Age', 'Fare']]
y = titantic_df[['Survived']]

In [46]:
X.head()

Pclass     Sex   Age     Fare
0       3    male  22.0   7.2500
1       1  female  38.0  71.2833
2       3  female  26.0   7.9250
3       1  female  35.0  53.1000
4       3    male  35.0   8.0500

In [47]:
y.head()

Survived
0         0
1         1
2         1
3         1
4         0

In [48]:
# converting to numerical values for column 'Sex'

from sklearn.preprocessing import LabelEncoder

In [49]:
le_sex = LabelEncoder()
X['sex_numerical'] = le_sex.fit_transform(X['Sex'])
X

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Pclass     Sex   Age     Fare  sex_numerical
0         3    male  22.0   7.2500              1
1         1  female  38.0  71.2833              0
2         3  female  26.0   7.9250              0
3         1  female  35.0  53.1000              0
4         3    male  35.0   8.0500              1
..      ...     ...   ...      ...            ...
886       2    male  27.0  13.0000              1
887       1  female  19.0  30.0000              0
888       3  female   NaN  23.4500              0
889       1    male  26.0  30.0000              1
890       3    male  32.0   7.7500              1

[891 rows x 5 columns]

In [50]:
# dropping sex as it is not required

X = X.drop(['Sex'],axis='columns')
X

Pclass   Age     Fare  sex_numerical
0         3  22.0   7.2500              1
1         1  38.0  71.2833              0
2         3  26.0   7.9250              0
3         1  35.0  53.1000              0
4         3  35.0   8.0500              1
..      ...   ...      ...            ...
886       2  27.0  13.0000              1
887       1  19.0  30.0000              0
888       3   NaN  23.4500              0
889       1  26.0  30.0000              1
890       3  32.0   7.7500              1

[891 rows x 4 columns]

In [51]:
y.head()

Survived
0         0
1         1
2         1
3         1
4         0

In [52]:
X.Age.isnull().sum()

177

We can observe that age has 177 NaN values

In [53]:
# Data Preprocessing: filling the NA value with median value of a column (Age)
import math
age = math.floor(X['Age'].median())
age

28

In [54]:
# assinging the value
X['Age'] = X['Age'].fillna(age)
X.Age.isnull().sum()

0

In [55]:
# import tree
from sklearn import tree
decision_model = tree.DecisionTreeClassifier()

In [56]:
# training the model
decision_model.fit(X,y)

DecisionTreeClassifier()

In [57]:
# accuracy of the model
decision_model.score(X,y)

0.9775533108866442

We have got an accuracy of 97%.

In [58]:
X.head()

Pclass   Age     Fare  sex_numerical
0       3  22.0   7.2500              1
1       1  38.0  71.2833              0
2       3  26.0   7.9250              0
3       1  35.0  53.1000              0
4       3  35.0   8.0500              1

In [59]:
# predicting if the person survived or not
decision_model.predict([[1,38.0,71.2833,0]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  "X does not have valid feature names, but"


array([1])

In [60]:
# predicting if the person survived or not
decision_model.predict([[3,35.0,8.0500,1]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  "X does not have valid feature names, but"


array([0])